docs/design/2024-09-06-maintain-priority-queue-in-memory.md

```diff
@@ -0,0 +1,567 @@
+        return err
+    }
+
+    event.ProcessedBy |= 1 << uint(handlerID)
+
+    return nil
+}
+func (n *DDLNotifier) fetchEvents() ([]DDLEvent, error) {
+    // ...
+    return nil, nil
+}
+
+func (n *DDLNotifier) hasProcessed(event DDLEvent, handlerID int) bool {
+    return (event.ProcessedBy & (1 << handlerID)) != 0
+}
+
+func (n *DDLNotifier) markProcessed(sctx sessionctx.Context, eventID int, handlerID int) error {
+    _, err := ExecRows(sctx, "UPDATE mysql.ddl_notifier SET processed_by = processed_by | (1 << ?) WHERE id = ?", handlerID, eventID)
+    return err
+}
+
+func (n *DDLNotifier) allHandlersProcessed(event DDLEvent) bool {
+    return event.ProcessedBy == (1<<len(n.handlers))-1
+}
+
+func (n *DDLNotifier) deleteProcessedEvent(sctx sessionctx.Context, eventID int) error {
+    _, err := ExecRows(sctx, "DELETE FROM mysql.ddl_notifier WHERE job_id = ?", eventID)
+    return err
+}
+
+func (n *DDLNotifier) RegisterHandler(id int, handler SchemaChangeHandler) {
+    n.handlers[id] = handler
+}
+```
+
+This notifier is responsible for fetching events from the `mysql.ddl_notifier` table, ordered by job ID. After it calls every handler for each event, it can safely delete handled events.
+We use a separate TXN to handle each handler. This would help us reduce the impact between each handler. After all handlers have handled this event, we will delete it.
+
+##### Where does it run?
+
+Since only the stats module is currently subscribed and requires in-memory operations, we will only start it at the stats owner.
+
+##### Future possibility
+
+If we do want other modules to subscribe to the module in the future, we need to avoid it only running in the stats owner node. This is because other modules may also have memory operations that may not occur in the stats owner node.
+**A possible solution would be to start DDLNotifier on each TiDB node, with different subscribers on each node. This would allow us to solve the problem.**
+However, this solution introduces more complexities, such as how to manage subscribers, and ensuring that an event can be deleted after it has been processed by subscribers on all the different nodes becomes more complicated. Even the granularity of the subscribers needs to be controlled. For example, when the same subscriber logic is used, but they are running on every node, whether we consider it to be a single subscriber or multiple subscribers is something that needs to be discussed.
+**Since this issue is already far from our goal this time, I decided to consider it as a future possibility rather than in the pursuit of completing it in this design.**
+
+#### Reflect on Time Change
+
+We need to consider a time factor here because one of the metrics used to calculate our scores is the analysis interval, which is the time interval between the last analysis and the present. Obviously, it changes over time and we need to keep recalculating the analysis interval for these tables.
```

Comment:
I think here you want to mention dynamic polling interval? Can you explain how we should change it? For exmaple, when the heap gets larger we should use smaller interval

Reply:
No, this is specifically for the Analysis Interval indicator. Time passes, so we need to recalculate the time duration based on the latest start timestamp from the system. However, because we want to keep it in memory, we need to update it periodically with the new stats timestamp.

Because it consists of all memory operations, I suggest using a fixed interval to refresh it, for example: every 5 minutes. I believe that would be acceptable.